pkg/registry/rbac/validation/kcp.go

```diff
@@ -0,0 +1,83 @@
+package validation
+
+import (
+	"context"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	authserviceaccount "k8s.io/apiserver/pkg/authentication/serviceaccount"
+	"k8s.io/apiserver/pkg/authentication/user"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	// ScopeExtraKey is the key used in a user's "extra" to specify
+	// that the user is restricted to a given scope. Valid values are:
+	// - "cluster:<name>"
+	// In the future, we might add:
+	// - "interval:from:to".
+	// Scopes are and'ed. Scoping to multiple clusters invalidates it for all.
+	ScopeExtraKey = "authentication.kcp.io/scopes"
+)
+
+type appliesToUserFunc func(user user.Info, subject rbacv1.Subject, namespace string) bool
+type appliesToUserFuncCtx func(ctx context.Context, user user.Info, subject rbacv1.Subject, namespace string) bool
+
+var appliesToUserWithWarrants = withScopes(appliesToUser)
+
+// withScopes wraps the appliesToUser predicate to check for the base user and any warrants.
+func withScopes(appliesToUser appliesToUserFunc) appliesToUserFuncCtx {
+	var recursive appliesToUserFuncCtx
+	recursive = func(ctx context.Context, u user.Info, bindingSubject rbacv1.Subject, namespace string) bool {
+		var clusterName logicalcluster.Name
+		if cluster := genericapirequest.ClusterFrom(ctx); cluster != nil {
+			clusterName = cluster.Name
+		}
```

Comment:
Dont we need to check for `cluster == ""` here? else its `""` and InSocpe function might get confusing? Mabe check inside `InScope` and if `""` just return false?

Reply:
`system:admin` logical cluster is empty string.